### PR TITLE
Serial request

### PR DESCRIPTION
**Pull request by iurly**
Here goes my proposal for a serial request (initially serial inout) node.
See the [original motivation](https://groups.google.com/d/topic/node-red/38aw5MkRRwk/discussion)   and the [recent dicussion](https://groups.google.com/forum/#!topic/node-red/mxki-T6pArE) on the mailing list.
Some notes:
- It shares the same configuration with `serial in` and `serial out` for any given port. However it should not affect the behavior of the existing nodes. This means you can use as many `serial in` nodes as you want along with the new node to "sniff" on the existing serial line and `serial out` to inject mailicious/tampering data onto the serial line skpping the queueing mechanism.
- I'm not sure the approach for duplicating a message is correct (Object.assign()), perhaps I'm potentially creating duplicate _msgid?
- I'm keeping a reference to the node (I mean Node-RED's node) within the internal queue to track which node the request originally came from, and hack the emitter paradigm to perform some filtering, I'm not sure this is 100% safe.
- ~most of the code is a copy from both serial in serial out, so I'm well aware it could use some refactoring -- feedback welcome!~ _recently refactored_
- every incoming message should eventually produce an outgoing message, I came up with the "status" field using some strings, perhaps there's a better solution
- this implementation shows some status on the node, which might turn annoying to some people. If accepted, it could use some improvement in that it does not distinguish between "message enqueued (X messages waiting)" and "message sent"
- ~ideally the timeout should also be configurable through the configuration node~ _DONE_.
- ~some other potential features (like delayed transmission) are stricken out from the documentation as they're not implemented~ _REMOVED from the documentation_.

**New feature**: as part of this PR there's also a new msg split logic: "silent line timeout". In this case the timeout is re-started at each received character.
Any feedback welcome! Thanks in advance!

### CODE REVIEW

File: io/serialport/25-serial.html [hunk 1]
```diff
@@ -82,6 +82,66 @@
...
+    with which it shares the configuration.</p>
+    <p>Send the request message in <code>msg.payload</code> as you would do with a <code>serial out</code> node.
+    The message will be forwarded to the serial port following a strict FIFO (First In, First Out) queue, waiting for a single response before transmitting the next request.
+    Once a response is received (with the same logic of a <code>serial in</code> node), or after a timeout occurs, a message is produced to the output (see Outputs below),
+    with <code>msg.payload</code> containing the received response (or missing in case if timeout) and all other fields preserved.</p>
+    <h3>Inputs</h3>
+    <ul>
+    <li>
+    <code>msg.timeout</code> is the timeout (in ms) after which the incoming message is propagated to the output with <code>msg.status</code> set to <code>"ERR_TIMEOUT"</code> and missing payload.
+    If not present, the default value is 10.000 (10s). <del>If set to zero, the message is immediately propagated to the output with <code>msg.status</code> set to <code>"OK"</code> and missing payload</del>.
```
Comment (dceejay) [inline io/serialport/25-serial.html]: these <del> tags render as strikethrough....  not needed.

File: io/serialport/25-serial.html [hunk 2]
```diff
@@ -82,6 +82,66 @@
...
+    </li>
+    <li><del><code>msg.delay</code> is the delay to wait before the message is sent to the serial device.
```
Comment (dceejay) [inline io/serialport/25-serial.html]: same again....

File: io/serialport/25-serial.html [hunk 3]
```diff
@@ -82,6 +82,66 @@
...
+    Default value is 0.</del></li>
+    <li><del>Set <code>msg.reset</code> to `true` to clean up the existing queue (in this case <code>msg.payload</code>
```
Comment (dceejay) [inline io/serialport/25-serial.html]: and again

**Comment by dceejay** [comment]
Hi @iurly - any chance to fix this up so we can merge it ?

**Comment by iurly** [comment]
Busy week, sorry! I'll pick this up again early next week. While at it, I was wondering if the unbounded queue is really a good idea and if we should do something about it... Thoughts?

**Comment by dceejay** [comment]
Indeed (re queue) - other nodes do allow you to specify depth in settings.js if you want (but don't require that you do). eg split, sort etc - maybe something like `serialRequestBufferMaxLength`

**Comment by iurly** [comment]
Any idea why my build is suddenly failing?

**Comment by dceejay** [comment]
So ... finally get a chance to look - seem good so far.
couple of nits to clean up..
image
should be made wider so the text fits.

Also the input node sets msg.port so you know where it comes from... is this the case here ? (though I guess as it's a request you know where you sent it so shouldn't need to know.)

The node.status messages should be translatable - using the locales/en-US/messages.json file rather than being hardcoded.

But yes - so far seems good.

File: io/serialport/25-serial.js [hunk 4]
```diff
@@ -358,7 +358,13 @@ module.exports = function(RED) {
                                 // do the timer thing
                                 if (spliton === "time") {
                                     // start the timeout at the first character
-                                    if (!obj.tout) {
+                                    if ( i == 1 ) {
```
Comment (dceejay) [inline io/serialport/25-serial.js]: This mode should only be enabled when in split on time from last character mode - not the current timeout mode.

Comment (iurly) [reply io/serialport/25-serial.js]: Hmm... keep in mind this version does not include the "new" behavior.
It behaves exactly as before: it starts the timeout the at the first character. It just uses a different test to establish it's the first character, as obj.tout could have been set for a different purpose already.
Anyway, I'm about to push the commit for the new behavior.

**Comment by dceejay** [comment]
Looks good
any further thoughts/changes/tidying before we final review / merge ?

**Comment by dceejay** [comment]
HI - any further thoughts/changes/tidying before we final review / merge ?

**Comment by dceejay** [comment]
I think this is the most "official" list - https://nodejs.org/api/errors.html#nodejs-error-codes
but I don't think they are sacred if there isn't one close enough.

**Comment by iurly** [comment]
I see, pretty interesting reading.
But I was talking about some standard way to notify of errors through `msg`s within Node-RED, if any.
`http request` reports some `msg.statusCode`, for instance (numeric, I guess).
I guess we may stick with the existing implementation then, it shouldn't sound totally insane.

**Comment by dceejay** [comment]
ah right. yes. the "normal" Node_RED way is to pass errors through node.error call instead - but again that's not what you want. I going numeric you could follow http... then again bash just goes 0 for  and anything else for a return code depending on severity... so ...
So yes leave it as-is I think.

There is an outstanding work-item on the backlist for how to signal success more generally (for nodes that have no output).

**Comment by dceejay** [comment]
OK - all merged. I'm going to leave it here for a week or so so we can more thoroughly test it before pushing to npm.

**Comment by iurly** [comment]
Cool! :-)
Do you have any instructions on how to install it for anyone who'd like to try it out?

**Comment by dceejay** [comment]
OK - I have created a beta tag so 

    cd ~/.node-red
    npm i node-red-node-serialport@beta

should do it

**Comment by yfhaotian** [comment]
Hello, do you know how serial request uses msg.baudrate to configure the baud rate?

**Pull request merged**